blockchain/utxoviewpoint.go

```diff
@@ -556,72 +575,74 @@ func (view *UtxoViewpoint) fetchInputUtxos(db database.DB, block *btcutil.Block)
 
 			// Don't request entries that are already in the view
 			// from the database.
-			if _, ok := view.entries[*originHash]; ok {
+			if _, ok := view.entries[txIn.PreviousOutPoint]; ok {
 				continue
 			}
 
-			txNeededSet[*originHash] = struct{}{}
+			neededSet[txIn.PreviousOutPoint] = struct{}{}
 		}
 	}
 
 	// Request the input utxos from the database.
-	return view.fetchUtxosMain(db, txNeededSet)
+	return view.fetchUtxosMain(db, neededSet)
 }
 
 // NewUtxoViewpoint returns a new empty unspent transaction output view.
 func NewUtxoViewpoint() *UtxoViewpoint {
 	return &UtxoViewpoint{
-		entries: make(map[chainhash.Hash]*UtxoEntry),
+		entries: make(map[wire.OutPoint]*UtxoEntry),
 	}
 }
 
-// FetchUtxoView loads utxo details about the input transactions referenced by
+// FetchUtxoView loads unspent transaction outputs for the inputs referenced by
 // the passed transaction from the point of view of the end of the main chain.
-// It also attempts to fetch the utxo details for the transaction itself so the
-// returned view can be examined for duplicate unspent transaction outputs.
+// It also attempts to fetch the utxos for the outputs of the transaction itself
+// so the returned view can be examined for duplicate transactions.
 //
 // This function is safe for concurrent access however the returned view is NOT.
 func (b *BlockChain) FetchUtxoView(tx *btcutil.Tx) (*UtxoViewpoint, error) {
-	b.chainLock.RLock()
-	defer b.chainLock.RUnlock()
-
-	// Create a set of needed transactions based on those referenced by the
-	// inputs of the passed transaction.  Also, add the passed transaction
-	// itself as a way for the caller to detect duplicates that are not
-	// fully spent.
-	txNeededSet := make(map[chainhash.Hash]struct{})
-	txNeededSet[*tx.Hash()] = struct{}{}
+	// Create a set of needed outputs based on those referenced by the
+	// inputs of the passed transaction and the outputs of the transaction
+	// itself.
+	neededSet := make(map[wire.OutPoint]struct{})
+	prevOut := wire.OutPoint{Hash: *tx.Hash()}
+	for txOutIdx := range tx.MsgTx().TxOut {
+		prevOut.Index = uint32(txOutIdx)
+		neededSet[prevOut] = struct{}{}
+	}
 	if !IsCoinBase(tx) {
 		for _, txIn := range tx.MsgTx().TxIn {
-			txNeededSet[txIn.PreviousOutPoint.Hash] = struct{}{}
+			neededSet[txIn.PreviousOutPoint] = struct{}{}
 		}
 	}
 
 	// Request the utxos from the point of view of the end of the main
 	// chain.
 	view := NewUtxoViewpoint()
-	err := view.fetchUtxosMain(b.db, txNeededSet)
+	b.chainLock.RLock()
```

Comment:
Why do you need this lock?

Reply:
It is doubling as a "block is currently be processed lock".

One of the things I think we need to do in the future is now that there is a lot more fine grained locking (which has been progressively happening for a while now), the overall chain lock will need to be replaced with a separate lock specifically for when a block is being processed.  There are places, such as from the RPC server and peer requests, that need to interrogate chain, but don't really care if a block is currently being processed.  However, the mining code very much cares if a block is being processed because it doesn't want to start work on a tip block that is in the process of being replaced, and likely, the mempool code cares since it doesn't want to process a transaction that is about to be invalid.  This code path is called from those places.

Reply:
But what does `fetchUtxosMain` need that is protected by the lock? Shouldn't database snapshotting make it safe?

Reply:
The fetch itself is safe without the lock from a pure concurrency perspective.  However, the intent of the lock is, as mentioned above, to ensure that callers are prevented from fetching the requested utxos while a block is currently being processed (which will update the database and thus change the utxos available to be fetched after it completes).

Scenario:

A new block shows up and is currently being processed via `ProcessBlock`.  The first thing that function does is take the chain lock.  It isn't released until the block is fully processed and hence the database is updated.

Simultaneously a new block template is requested which requires looking up the current utxos to check the generated block is valid (aside from proof of work) before delivering it to the miner to begin work on it.  Naturally, it's building the block on the one before the block that is currently being processed, because as far as it knows, the current best block will remain the present value (current block could fail to process).

So, by waiting here, when the block finishes processing, it means the calling code will have a view that no longer matches the best block which will trigger it to build a new template.  Without it, it would simply succeed and return a block building on the old tip.

As I mentioned before, this could definitely be made more explicit and clear by using a separate lock and is something that should ultimately happen.  I 100% agree it's not very intuitive.  It is really all just a relic of the fact that _everything_ (except orphans) in `blockchain` used to have a single lock and the rest of the aforementioned code relies on that still.  The code has been incrementally introducing more fine-grained locking.